src/utils.rs

```diff
@@ -0,0 +1,158 @@
+use flate2::read::GzDecoder;
+use std::path::Path;
+use std::process::{Command, ExitStatus};
+use std::{fs, io};
+use tar::Archive;
+
+/// Filenames that if found in a project mean it should be treated as a Python project,
+/// and so pass this buildpack's detection phase.
+///
+/// This list is deliberately larger than just the list of supported package manager files,
+/// so that Python projects that are missing some of the required files still pass detection,
+/// allowing us to show a more detailed error message during the build phase than is possible
+/// during detect.
+const KNOWN_PYTHON_PROJECT_FILES: [&str; 9] = [
+    ".python-version",
+    "main.py",
+    "manage.py",
+    "Pipfile",
+    "poetry.lock",
+    "pyproject.toml",
+    "requirements.txt",
+    "runtime.txt",
+    "setup.py",
+];
+
+/// Returns whether the specified project directory is that of a Python project, and so
+/// should pass buildpack detection.
+pub(crate) fn is_python_project(app_dir: &Path) -> io::Result<bool> {
+    // Until `Iterator::try_find` is stabilised, this is cleaner as a for loop.
+    for filename in KNOWN_PYTHON_PROJECT_FILES {
+        if app_dir.join(filename).try_exists()? {
+            return Ok(true);
+        }
+    }
+
+    Ok(false)
+}
+
+/// Read the contents of the provided filepath if the file exists, gracefully handling
+/// the file not being present, but still returning any other form of IO error.
+pub(crate) fn read_optional_file(path: &Path) -> io::Result<Option<String>> {
+    fs::read_to_string(path)
+        .map(Some)
+        .or_else(|io_error| match io_error.kind() {
+            io::ErrorKind::NotFound => Ok(None),
+            _ => Err(io_error),
+        })
+}
+
+/// Download a gzipped tar file and unpack it to the specified directory.
+pub(crate) fn download_and_unpack_gzipped_archive(
+    uri: &str,
+    destination: &Path,
+) -> Result<(), DownloadUnpackArchiveError> {
+    // TODO: (W-12613141) Add a timeout: https://docs.rs/ureq/latest/ureq/struct.AgentBuilder.html?search=timeout
+    // TODO: (W-12613168) Add retries for certain failure modes, eg: https://github.com/algesten/ureq/blob/05b9a82a380af013338c4f42045811fc15689a6b/src/error.rs#L39-L63
```

Comment:
While you are adding TODOs - maybe one to verify the download against a known checksum? This might be more of an epic though, considering the existing comments around adding and using an inventory/manifiest.

Reply:
Yeah I have this epic for that:
[Python CNB: Validate the SHA256 checksum of Python binaries](https://gus.lightning.force.com/lightning/r/ADM_Epic__c/a3QEE000000V4vN2AS/view)